### PR TITLE
Epetra #480 setup for the LAPACK testing of the dggsvd3 and sggsvd3.

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
Adds configure time check for dggsvd3 through CheckHaveLapackGSSVD3.cmake.
Then defines HAVE_LAPACK_GSSVD3 which adds the new parameter LWORK to dggsvd3 and sggsvd3.

This was tested on LAPACK 3.6.0.

To resolve issue #480.

### CODE REVIEW

**Comment by bmpersc** [comment]
The cmake portions look good to me. The only concern I have is searching for multiple manglings of the name DGGSVD3. But I don't see a good way to figure out the Fortran mangling that wouldn't put an unnecessary burden on a lot of legitimate use cases that don't need a Fortran compiler. I think the manglings that are checked are the likely ones to occur so this should be fine for a while. Once this has been pushed can you file a ticket against Epetra noting why the Fortran mangling wasn't used so that we have a record of why the decision was made and a reminder to fix it if we ever find a good solution to this problem.

**Comment by jwillenbring** [comment]
@mhoemmen could you review the implementation of the pull request? @bmpersc has reviewed the cmake aspects.

File: packages/epetra/cmake/CheckHaveLapackGSSVD3.cmake [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#   Information is at:
+#   http://www.netlib.org/lapack/explore-html/d6/db3/dggsvd3_8f_a4a187519e5c71da3b3f67c85e9baf0f2.html#a4a187519e5c71da3b3f67c85e9baf0f2
+
+INCLUDE(CheckFunctionExists)
+
+FUNCTION(CHECK_HAVE_LAPACK_GSSVD3 VARNAME)
+
+  SET(CMAKE_REQUIRED_LIBRARIES ${TPL_LAPACK_LIBRARIES})
+
+  CHECK_FUNCTION_EXISTS( dggsvd3   HAVE_dggsvd3 )
```
Comment (mhoemmen) [inline packages/epetra/cmake/CheckHaveLapackGSSVD3.cmake]: Lines 21-24 limit the set of possible Fortran manglings.  Why not just use the existing CMake FortranCInterface (de)mangling functions?

**Comment by mhoemmen** [comment]
@jwillenbring I'm working on this one now -- sorry didn't see this before.

File: packages/epetra/cmake/CheckHaveLapackGSSVD3.cmake (hunk 1 above)
Comment (krcb) [inline packages/epetra/cmake/CheckHaveLapackGSSVD3.cmake]: See the comment from @bmpersc above.

File: packages/epetra/cmake/Epetra_config.h.in [hunk 2]
```diff
@@ -110,6 +110,9 @@
 /* Define if you have LAPACK library. */
 #cmakedefine HAVE_LAPACK
 
+/* Define if you have new LAPACK dgssvd3 and fgssvd3 functions which added a new parameter */
+#cmakedefine HAVE_LAPACK_GSSVD3
```
Comment (mhoemmen) [inline packages/epetra/cmake/Epetra_config.h.in]: I'm not keen on Epetra's highly generic macro names.  Would prefer prefixing in a way that's consistent with other Trilinos packages: `HAVE_EPETRA_LAPACK_GSSVD3`.

File: packages/epetra/src/CMakeLists.txt [hunk 3]
```diff
@@ -16,6 +16,10 @@ SET(HAVE_BLAS ON)
 SET(HAVE_LAPACK ON)
 SET(HAVE_MPI ${Epetra_ENABLE_MPI})
 
+INCLUDE(CheckHaveLapackGSSVD3)
+CHECK_HAVE_LAPACK_GSSVD3(HAVE_LAPACK_GSSVD3)
```
Comment (mhoemmen) [inline packages/epetra/src/CMakeLists.txt]: See above comment on Epetra's macros not being namespaced.

**Comment by MicheldeMessieres** [comment]
@mhoemmen  
I added Epetra name space labels as suggested. Also had found the BLAS TPL path should be specified in the cmake file so have changed that. I fixed the pull branch to be against develop instead of master. So this has been tested on old and new versions of Lapack for both linux and mac builds, though I did not specifically try a Lapack 3.1.1. yet.

**Comment by mhoemmen** [comment]
> So this has been tested on old and new versions of Lapack for both linux and mac builds, though I did not specifically try a Lapack 3.1.1. yet.

Could you clarify what's an "old" version?  Does "old" mean "before this interface changed"?
If so, then we don't need to try LAPACK 3.1.1.  I only mention that version because it's (a) a common default on older systems, and (b) there were lots of interface changes between 3.1.1 and 3.2.2.

**Comment by MicheldeMessieres** [comment]
To clarify, by old I meant Lapack which doesn't have the new dggsvd3 function. For Linux I had tried Lapack 3.2.1 which does not detect dggsvd3, and Lapack 3.6.0 which does detect dggsvd3.

**Comment by mhoemmen** [comment]
> To clarify, by old I meant Lapack which doesn't have the new dggsvd3 function. For Linux I had tried Lapack 3.2.1 which does not detect dggsvd3, and Lapack 3.6.0 which does detect dggsvd3.

Sounds good then -- thanks for clarifying!

**Comment by mhoemmen** [comment]
@jwillenbring I'm done reviewing :-)

**Comment by krcb** [comment]
@jwillenbring can this be merged?

**Comment by mhoemmen** [comment]
@krcb Does it still pass tests?

**Comment by krcb** [comment]
@MicheldeMessieres is verifying now and will update.

**Comment by MicheldeMessieres** [comment]
Rebased to latest develop and verified this on linux (serial and MPI) with Epetra enabled for checkin script but not with forward packages enabled. Also validated on a Mac. Will take me a while to get the forward packages also tested if we need that - having some trouble with the checkin script. But Epetra tests pass in above cases. (Though note this method dggsvd is not used in any of the current code base I have so this is just verifying the API is setup to compile properly).

**Comment by mhoemmen** [comment]
Thanks for checking!  I just merged in your changes.

**Pull request merged**